kdc/kerberos5.c

```diff
@@ -1819,12 +1819,31 @@ static krb5_error_code
 generate_pac(astgs_request_t r, const Key *skey, const Key *tkey)
 {
     krb5_error_code ret;
+    const krb5_keyblock *pk_reply_key = NULL;
     krb5_pac p = NULL;
     krb5_data data;
     uint16_t rodc_id;
     krb5_principal client;
+    krb5_boolean sent_pac_request;
+    krb5_boolean pac_request;
+
+    ret = check_pa_pac_request(r->context, &r->req, &pac_request);
+
+    sent_pac_request = (ret == 0);
+
+    switch (r->validated_pa_type) {
+    case KRB5_PADATA_PK_AS_REQ:
+    case KRB5_PADATA_PK_AS_REQ_WIN:
+	pk_reply_key = &r->reply_key;
+	break;
+    }
```

Comment:
This is for including NTLM keys when the client doesn't authenticate with long-term keys, right? In this case I think we actually want a flag on the PA mechanism that indicates a long-term key was used, e.g. `PA_LONG_TERM_KEY`. This should be set on `ENC-TS` and `ENC-CHAL`. Then lets store a flag in the request:

```c
unsigned int pa_validated_long_term_key : 1;
```

where `found_pa` would have been set, and pass `r->pa_validated_long_term_key ? NULL : &r->reply_key` as an argument to `_kdc_pac_generate()`.

Eliminate the temporary `pk_reply_key` variable and `validated_pa_type`.

Reply:
Or, if you want, invert the logic and make the flag `PA_REPLACE_REPLY_KEY` (on PKINIT and GSS), and propagate when validated to `r->pa_replaced_reply_key`. Then, if *that* is set, you ask the backend to include the symmetries keys.

Reply:
`PA_REPLACE_REPLY_KEY` is a little more semantically consistent with RFC6113.

Reply:
The variable is communicating two things:
 * the key to encrypt the NTLM hash with
 * that an NTLM hash should be returned

A better design might be to always pass in the reply_key and the PA data and let the HDB module choose what PA types should trigger the NTLM reply.  Would that be OK?

Reply:
I think either is fine. The advantage of doing it in the front end is that it's resilient to new PA types and, perhaps we want to flag which PA types replace the reply key anyway, for some other unspecified future reason. But I'm also inclined to take the patch that changes the fewest lines of code ;)

Reply:
It isn't replacing the reply key, it is using it.

Reply:
See [RFC 6113 Section 3.3](https://datatracker.ietf.org/doc/html/rfc6113.html#section-3.3) and [RFC 4456 Section 3.2.3.2](https://datatracker.ietf.org/doc/html/rfc4556#section-3.2.3.2).

PKINIT and GSS are examples of two pre-authentication mechanisms that replace the reply key.

Mechanisms that replace the reply key by definition do not encrypt the reply in the client's long-term key. For downlevel compatibility Windows needs to include the NTLM OWFs for NTLM compatibility (of course, this case the OWFs can be randomly generated, as the user may never have a password set).

Reply:
Thanks, now I get it.  Now I get why a flag up at the PA types level makes sense here.

Reply:
Another point just for posterity: the KDC is implicitly authenticated to the client when the reply key is *not* replaced, because it’s proved knowledge of the client’s long term key. When a pre-authentication mechanism replaces the reply key, the client needs another way of authenticating the KDC: validating its certificate, checking GSS_C_MUTUAL_FLAG, etc.

Reply:
Do you want me to prepare a patch for this based on the proposed solution?

Reply:
I have a patch, will file a separate PR